### PR TITLE
Refactor Kubelet to conform to the internal service manager

### LINKED ISSUE #201: Refactor kubelet to conform to common service interface
#### Design Document Link

PR: #217 
Parent issue https://github.com/redhat-et/microshift/issues/161

### PR DESCRIPTION
**Pull request by husky-parul**
**Which issue(s) this PR addresses**:
https://github.com/redhat-et/microshift/issues/201

Closes #201

### CODE REVIEW

File: pkg/node/kubelet.go [hunk 1]
```diff
@@ -47,17 +86,50 @@ func StartKubelet(cfg *config.MicroshiftConfig) error {
...
+	kubeletServer := &kubeletoptions.KubeletServer{
+		KubeletFlags:         *s.kubeletflags,
+		KubeletConfiguration: *s.kubeconfig,
+	}
+
+	kubeletDeps, err := kubelet.UnsecuredDependencies(kubeletServer, utilfeature.DefaultFeatureGate)
+	if err != nil {
+		logrus.Fatalf("Error in fetching depenedencies %v", err)
+	}
+	if err := kubelet.Run(ctx, kubeletServer, kubeletDeps, utilfeature.DefaultFeatureGate); err != nil {
```
Comment (rootfs) [inline pkg/node/kubelet.go:128]: add a readiness probe (similar to kube-controller-manger or kube-scheduler), using kubelet healthz option `healthzPort` https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/

File: pkg/node/kubelet.go [hunk 2]
```diff
@@ -47,17 +86,50 @@ func StartKubelet(cfg *config.MicroshiftConfig) error {
...
+	kubeletoptions.AddGlobalFlags(cleanFlagSet)
 
+	if err := cmd.ParseFlags(args); err != nil {
+		logrus.Fatalf("failed to parse flags:%v", err)
+		return err
+	}
+	s.kubeletflags = kubeletFlags
+	s.kubeconfig = kubeletConfig
+	s.kubeconfigfile = filepath.Join(cfg.DataDir, "resources", "kubelet", "kubeconfig")
+	logrus.Infof("Starting kubelet %s, args: %v", cfg.HostIP, args)
```
Comment (rootfs) [inline pkg/node/kubelet.go:110]: HostIP -> NodeIP

Comment (husky-parul) [reply pkg/node/kubelet.go]: It created a merge conflict. I have rebased the PR branch

File: pkg/node/kubelet.go [hunk 3]
```diff
@@ -47,19 +87,51 @@ func StartKubelet(cfg *config.MicroshiftConfig) error {
...
+	kubeletoptions.AddKubeletConfigFlags(cleanFlagSet, kubeletConfig)
+	kubeletoptions.AddGlobalFlags(cleanFlagSet)
+
+	if err := cmd.ParseFlags(args); err != nil {
+		logrus.Fatalf("failed to parse flags:%v", err)
+		return err
+	}
+	s.kubeletflags = kubeletFlags
+	s.kubeconfig = kubeletConfig
+	s.kubeconfigfile = filepath.Join(cfg.DataDir, "resources", "kubelet", "kubeconfig")
```
Comment (rootfs) [inline pkg/node/kubelet.go:110]: is `kubeconfigfile` used after assignment?

Comment (husky-parul) [reply pkg/node/kubelet.go:110]: It's not.

File: pkg/node/kubelet.go [hunk 4]
```diff
@@ -116,6 +116,17 @@ func (s *Kubelet) configure(cfg *config.MicroshiftConfig) error {
 func (s *Kubelet) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
 	defer close(stopped)
 
+	// run readiness check
+	go func() {
+		healthcheckStatus := util.RetryInsecureHttpsGet("https://127.0.0.1:10248/healthz")
+		if healthcheckStatus != 200 {
+			logrus.Fatalf("Kube-controller-manager failed to starttttttt")
```
Comment (oglok) [inline pkg/node/kubelet.go:123]: Kubelet instead of kube-ctrl-mgr  and remove repeated "t"s :-)

File: pkg/node/kubelet.go [hunk 5]
```diff
@@ -116,6 +116,17 @@ func (s *Kubelet) configure(cfg *config.MicroshiftConfig) error {
...
+		}
+
+		logrus.Infof("%s is readyyy", s.Name())
```
Comment (oglok) [inline pkg/node/kubelet.go:126]: Remove repeated "y"s

File: pkg/node/kubelet.go [hunk 6]
```diff
@@ -16,19 +16,59 @@ limitations under the License.
...
 	kubelet "k8s.io/kubernetes/cmd/kubelet/app"
+	kubeletoptions "k8s.io/kubernetes/cmd/kubelet/app/options"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
 )
 
-func StartKubelet(cfg *config.MicroshiftConfig) error {
-	command := kubelet.NewKubeletCommand()
+const (
+	// Kubelet component name
+	componentKubelet = "kubelet"
```
Comment (oglok) [inline pkg/node/kubelet.go:41]: For consistency, either we use the const in all places (like L57), or we just remove the const and use the string "kubelet".

File: pkg/node/kubelet.go (hunk 5 above)
Comment (husky-parul) [reply pkg/node/kubelet.go:126]: Sorry I added this for easy grep search in microshift logs for the sake of debugging. The e2e are failing and I was trying to debug it. I will remove it.

**Comment by oglok** [comment]
Two comments on this PR.

- Testing your PR, It seems to be failing in my env:

`FATA[0011] Kubelet failed to start failed to run Kubelet: running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false. /proc/swaps contained: [Filename                    Type             Size            Used            Priority /dev/nvme0n1p6                          partition      8202236         709804          -2] `

-  Also, Kustomizer was moved to a different place.

**Comment by oglok** [comment]
> Two comments on this PR.
> 
> * Testing your PR, It seems to be failing in my env:
> 
> `FATA[0011] Kubelet failed to start failed to run Kubelet: running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false. /proc/swaps contained: [Filename Type Size Used Priority /dev/nvme0n1p6 partition 8202236 709804 -2] `
> 
> * Also, Kustomizer was moved to a different place.

Disabling swapping gets it to work. However, I don't know if that's a requirement for Kubelet. wdyt @fzdarsky ?

I also noticed a couple of deprecation warnings that we might want to address now:

```
Flag --cgroup-driver has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.                                                                                                                                                                                         Flag --volume-plugin-dir has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.      
```

**Comment by fzdarsky** [comment]
> Disabling swapping gets it to work. However, I don't know if that's a requirement for Kubelet. wdyt @fzdarsky ?

Current best practice is to ensure swap is off on the host, as swap may lead to various resource management issues in kubelet. However, I don't think we can / want to generally assume this, e.g. on a developer's machine. Also, I understand swap will be supported down the line (see https://kubernetes.io/blog/2021/08/09/run-nodes-with-swap-alpha/).

To not fail on swap being enabled, the `--fail-swap-on=false` flag can be used I believe.

**Comment by fzdarsky** [comment]
> I also noticed a couple of deprecation warnings that we might want to address now:
> 
> ```
> Flag --cgroup-driver has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
> 
> Flag --volume-plugin-dir has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.      
> ```

I agree it'd be great to fix this while we're at it.

File: pkg/node/kubelet.go (hunk 6 above)
Comment (mangelajo) [reply pkg/node/kubelet.go:41]: Another option is to use constants once you have more than one reference to the same string, that's what we did @ submariner-

File: pkg/node/kubelet.go [hunk 7]
```diff
@@ -16,26 +16,61 @@ limitations under the License.
...
+		logrus.Infof("Failed to create a new kubelet configuration: %v", err)
+		return err
+	}
+	// Prepare commandline args
 	args := []string{
 		"--config=" + cfg.DataDir + "/resources/kubelet/config/config.yaml",
 		"--bootstrap-kubeconfig=" + cfg.DataDir + "/resources/kubelet/kubeconfig",
 		"--kubeconfig=" + cfg.DataDir + "/resources/kubelet/kubeconfig",
 		"--container-runtime=remote",
 		"--container-runtime-endpoint=/var/run/crio/crio.sock",
```
Comment (rootfs) [inline pkg/node/kubelet.go:71]: per recent fix #248 this is now "unix://..."

File: pkg/node/kubelet.go [hunk 8]
```diff
@@ -16,26 +16,61 @@ limitations under the License.
...
 		"--runtime-cgroups=/system.slice/crio.service",
+		"--cgroup-driver=systemd",
```
Comment (rootfs) [inline pkg/node/kubelet.go:73]: this is already set [here](https://github.com/redhat-et/microshift/blob/main/pkg/config/kubelet.go#L37)

File: pkg/node/kubelet.go [hunk 9]
```diff
@@ -46,18 +81,67 @@ func StartKubelet(cfg *config.MicroshiftConfig) error {
...
+	cleanFlagSet := pflag.NewFlagSet(componentKubelet, pflag.ContinueOnError)
+	cleanFlagSet.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
 
-	if err := command.ParseFlags(args); err != nil {
-		logrus.Fatalf("failed to parse flags:%v", err)
+	kubeletFlags := kubeletoptions.NewKubeletFlags()
+	kubeletConfig, err := kubeletoptions.NewKubeletConfiguration()
+	// programmer error
+	if err != nil {
+		logrus.Fatalf("programmer error %v", err)
```
Comment (rootfs) [inline pkg/node/kubelet.go:91]: maybe a more informative error message?

**Comment by husky-parul** [comment]
> > Disabling swapping gets it to work. However, I don't know if that's a requirement for Kubelet. wdyt @fzdarsky ?
> 
> Current best practice is to ensure swap is off on the host, as swap may lead to various resource management issues in kubelet. However, I don't think we can / want to generally assume this, e.g. on a developer's machine. Also, I understand swap will be supported down the line (see https://kubernetes.io/blog/2021/08/09/run-nodes-with-swap-alpha/).
> 
> To not fail on swap being enabled, the `--fail-swap-on=false` flag can be used I believe.

It is set in [/pkg/config/kubelet.go/KubeletConfig]( https://github.com/redhat-et/microshift/blob/main/pkg/config/kubelet.go#L39)

**Pull request merged**